crates/compiler/gen_dev/src/generic64/mod.rs

```diff
@@ -746,6 +770,40 @@ impl<
                     .load_to_general_reg(&mut self.buf, src2);
                 ASM::imul_reg64_reg64_reg64(&mut self.buf, dst_reg, src1_reg, src2_reg);
             }
+            Layout::Builtin(Builtin::Int(
+                IntWidth::U64 | IntWidth::U32 | IntWidth::U16 | IntWidth::U8,
+            )) => {
+                // TODO find a general way to do this
+                let rax = CC::GENERAL_RETURN_REGS[0];
+                let rdx = CC::GENERAL_RETURN_REGS[1];
+
+                self.storage_manager
+                    .ensure_reg_free(&mut self.buf, RegStorage::General(rax));
+
+                self.storage_manager
+                    .ensure_reg_free(&mut self.buf, RegStorage::General(rdx));
+
+                self.storage_manager
+                    .load_to_specified_general_reg(&mut self.buf, src1, rax);
+
```

Comment:
so this clears rax and rdx, then loads src1 into rax (assuming rax will be the first argument and the general register the second, based on its "type" signature)

Reply:
So I think this will actually be the same as `imul` only in `x86_64.rs` will we handle this mess. This means we will have arbitrary input and output registers to begin with.

A concrete example:
1. call `umul` with src1 = `R9`, src2 = `R10`, dst = `R11`
2. free `RAX` and `RDX`
3. move `R9` to `RAX`
4. call underlying mul instructions with `R10`
5. move `RAX` to `R11`

Picking regs and 1 happens here.
2-5 happen in `x86_64.rs`
Of course we can do some optimizations if we happen to be putting the result in `RAX` or already using one of the regs.

Reply:
thanks, that was very helpful. But, it still does not seem to work. the numeric output is wrong (but constant), and the disassembler test just always gives an empty string.